File: Java/Medium/Accepted/verify-preorder-serialization-of-a-binary-tree_8.java
class Solution {
    public boolean isValidSerialization(String preorder) {
        
        int n = preorder.length();
        if (n == 0) return false;
        int i = 0;
        int slots = 1;
        while (i < n) {
            if (slots == 0) return false;
            if (preorder.charAt(i) == ',') {
                i++;
            } else if (preorder.charAt(i) == '#') {
                slots--;
                i++;
            } else {
                while (i < n && preorder.charAt(i) != ',') {
                    i++;
                }
                slots++;
            }
        }
        return slots == 0;
    }
}

File: Python3/Medium/WrongAnswer/verify-preorder-serialization-of-a-binary-tree_8.py
class Solution:
    def isValidSerialization(self, preorder: str) -> bool:
        
        preorder = preorder.split(',')
        stack = []
        for i, c in enumerate(preorder):
            if i == 0:
                stack.append(c)
            else:
                if c == '#' and stack[-1] == '#':
                    stack.pop()
                    if not stack:
                        return False
                    stack[-1] = '#'
                else:
                    stack.append(c)
        return stack == ['#']